### ISSUE TITLE
Static API for discovering events

### ISSUE THREAD

**Issue by josevalim** [beam-telemetry/telemetry#56]
_Originally posted by @bryannaegele in https://github.com/beam-telemetry/telemetry/issues/12#issuecomment-572889254_:

We've discussed a number of ways to discover telemetry events for a variety of use cases, most recently as it pertains to registering of event handlers for tracing purposes. Knowledge of event names is a predicate to registering handlers, so we are left with two potential avenues: compile-time introspection or explicit declaration.

Note that this is not intended to be the sole mechanism for event discovery. Issue #12 talks about a way to discover events as they happen. This discussion is about discovering events before they happen so they can be listened to in an unintrusive way.

### Compile-time Introspection

This is arguably the best overall solution since it requires no additional steps by the user. In my initial investigation, introspection would be difficult to achieve, especially across languages.

### Explicit Declaration

By requiring the user or library author to explicitly declare possible events, we have a straightforward mechanism to make use of that declaration at various steps and with a broad range of applications.

#### Behaviour

One option would be to provide a behaviour with a `telemetry_events/0` function which returns a list of discoverable events.

#### Module Attribute

Introduce an attribute of `telemetry_events` with a list of discoverable events.

Each of the above is easily discoverable at runtime and compile-time. My initial idea was to use module attributes but, after further consideration, I'm leaning towards @josevalim's idea of going with a behaviour. It's just as easy to use for a user, feels more idiomatic, and provides flexibility going forward. It's probably more useful from a tooling perspective, as well.

### Proposal

- [ ] Introduce a `telemetry` behaviour with a `telemetry_events/0` callback (default impl returns `[]`).
- [ ] Expand the API to add `telemetry:register_event/1`, `telemetry:deregister_event/1`, and `telemetry:register_all_events/1` functions to support further runtime declaration flexibility and discoverability (ETS storage is probably fine).
  - [ ] Fire events for invocation of each of the above which a tracing or other interested library could listen for.
  - [ ] Validate event name declarations during registration.
  - [ ] Validate against duplicate event names. (Raise an exception? This situation is likely to cause handler detachments. We should at least warn.)
  - [ ] Store registering module names with the event name? This could be useful in debugging or warning against duplicate event names being registered from different namespaces.
- [ ] Add `telemetry:list_all_events/0` would return all registered events.
- [ ] Introduce an event definition data structure for discovery and documentation. Includes event name, event description, available measurements with optional descriptions, metadata with optional descriptions.

/cc @arkgil @GregMefford @tsloughter @hauleth @binaryseed

**Comment by josevalim** [comment]
Perfect @bryannaegele! To add a couple more points that came up in the discussion:

 * Some questions were raised about the usefulness of simply listing all events that may happen, because without the metadata and particular information about the events, knowing only the event names themselves is not very useful.
  * One of the issues was discovery of Ecto events, which can be solved in Ecto specific ways. For example, querying Ecto about all running processes or traversing all loaded modules. This perhaps should be moved to Ecto issues tracker.
  * One of the issues was discovery of span events, which can be useful to know even without the metadata.

In a nutshell, my point is that I am not sure if we should have an API to discover all events, as it may have limited usability. Maybe we should focus on addressing the immediate issues: Ecto and spans. Can anyone add more scenarios were discoverability is necessary?

**Comment by hauleth** [comment]
I think that `telemetry:list_all_events/0` should be 2 functions:

- `telemetry:list_all_module_events/1`
- `telemetry:list_all_application_events/1`

To have finer graded access to the events as well.

BTW I think that module attribute has one advantage over behaviour/function - module attribute could be used by documentation generating tool or other "external tools" without executing module code. Maybe that would be more useful if we would to include the exposed events in the documentation as well.

**Comment by bryannaegele** [comment]
Good callout @hauleth on both points. I was thinking we could achieve both in the one function (check if atom is an application, then check if it's a module if it isn't found) but we could surface that more explicitly.

For usefulness of knowing all declared event names (there could be an argument made that you'd want an event to be private for some reason), I think we get a simple base mechanism for upfront and further discovery without resorting to tracing or compilation tactics. It would be possible to build on top of these APIs for a few purposes:

- Registering an event and tracking the module it was found in provides a mechanism to attach a handler to each instance to collect further metadata for your purpose when/if they execute. You could add the module name to the handler's config to know where it came from.
- Spans can be discovered by examining the registered events which could be spanned by reversing the event names and matching from there, e.g. `[:stop, :query, :repo]` having a matching `[:start, :query, :repo]` gives us a spannable set of events which we can create handlers for.
- Debug logging as mentioned in #12 is fairly simple to attach/detach handlers to dump everything from all events or a subset

Automatic documentation would be lacking without a more complex data structure with metadata for descriptions and such. We could build libraries on top of these APIs to make that feasible pretty easily, though.

**Comment by hauleth** [comment]
> I was thinking we could achieve both in the one function (check if atom is an application, then check if it's a module if it isn't found)

And then `:logger` would bring everything down as it is module in OTP 21+ `kernel` application and this is Elixir application name. There is no simple way to differentiate between them without user help.

> there could be an argument made that you'd want an event to be private for some reason

In that case I believe the `telemetry` is wrong solution. `telemetry` is just logging system without levels, filters, and all that stuff (however I think filters would be nice addition). You do not want to have "private" logs.

BTW I was even thinking about ditching `telemetry` in general and using structured logging of OTP to have similar functionality. However the problem was - which log level use so user would not accidentally disable logs for them. In contrast to `lager` Erlang `logger` has "default" level and nothing that is below it will not go further.

**Comment by josevalim** [comment]
> BTW I was even thinking about ditching telemetry in general and using structured logging of OTP to have similar functionality. However the problem was - which log level use so user would not accidentally disable logs for them. In contrast to lager Erlang logger has "default" level and nothing that is below it will not go further.

We had this discussion a couple times already, so we should stick the discussion somewhere so we don't revisit it every 2 or 3 months. There are reasons why telemetry works better than logger for what it intends to but I don't think we should side-track this issue by rehashing the same arguments.

**Comment by hauleth** [comment]
> There are reasons why telemetry works better than logger for what it intends

I know, and I have pointed them above. What was my intention there was to point that in the end Telemetry is still logging library, just with one specific purpose in mind.

**Comment by josevalim** [comment]
Apologies, I misread it then. :)

**Comment by binaryseed** [comment]
So the reason I've brought this up a few times is that it makes auto-instrumentation more difficult when you can't know upfront what the event name will be.

Ecto emits events like this:

```elixir
[:configurable, :prefix, :query]
```

So since it's fully configurable, I need to know that configuration before I can `attach` to the event.

If we re-think this just a bit, `ecto` could emit an event like:

```elixir
[:ecto, :query, :configurable, :suffix]
```

Then we could provide a way to specify a wildcard in our `attach`, perhaps:

```elixir
:telemetry.attach("handler", [:ecto, :query, :_], &handler/4)

def handler([:ecto, :query | _suffix], _, _, _) do
  # ...
end
```

With this idea, we provide customizable event names & provide a way to attach to all `ecto` events upfront w/o knowing details about the ecto repo configuration.

With that, the use case of auto-discovery would be solved, and a static API wouldn't be needed. We could still provide a runtime API for event discovery...

**Comment by binaryseed** [comment]
Another idea that would remove the need for knowing the event types upfront that would be a whole lot simpler: let a handler attach to _all_ events.

**Comment by josevalim** [comment]
@binaryseed but what can we do by subscribing to all events, if you don’t know what they mean and what is their metadata about? I understand if we were using them for spans, but not all of them are spans. As I asked above, is there any reason to add these features besides to solve the specific problems of spans and Ecto?

**Comment by binaryseed** [comment]
The same challenge will exist for Plug as well, and any library that goes with configurable event names. I think Tesla is going the configurable route as well in a recent PR.

For my use case of auto-instrumentation, the primary goal is ease of use for the end user. Subscribing to all events does mean that some would have no meaning and would effectively be a no-op. That’s fine, I’ll have function heads that handle all the events that the agent does know about and have logic for translating known metadata into metrics, spans, traces etc.

Besides ease of use, one of the motivations behind auto instrumentation is for situations where the library doing work is and indirect dependency and the developer of the app didn’t configure it. Expert devs will not have much issue in many of these cases but for many people they just need visibility into what the app is doing not another problem to solve. Tracking down and configuring every usage of telemetry is a meaningful barrier. It’s not insurmountable but it is real. What I hope to avoid is the situation where we have a lot of duplicated config since both telemetry and the instrumentation vendor need the same config values. If those get out of sync you loose visibility.

That said I can accept if we decide not to solve this directly. I can go the route of tracing functions to discover event names when the app is starting up. Simply having the libraries instrument themselves is already a far better situation than most other language ecosystems (no monkey patching or byte-code manipulation)!

**Comment by arkgil** [comment]
> Then we could provide a way to specify a wildcard in our attach

Unfortunately, this approach has been tested in the past and it was much less performant than matching exactly on the event name, as we have now. Note sure how big of a difference it was, I could try digging up some old benchmarks we did.

My take on whether the event names should be configurable is that no matter which route we take, there will be folks who would want it the other way. Some would like to differentiate distinct sources of events based on metadata, some would prefer to do it based on the event name. Given that, I suppose it's better to stick with what we already have, and assume that some event names will be configurable.

**Comment by arkgil** [comment]
> Spans can be discovered by examining the registered events which could be spanned by reversing the event names and matching from there, e.g. [:stop, :query, :repo] having a matching [:start, :query, :repo] gives us a spannable set of events which we can create handlers for.

I like it! With this approach and listing of event names per application we could enable/disable tracing on the per application basis.

**Comment by binaryseed** [comment]
>> Then we could provide a way to specify a wildcard in our attach

> Unfortunately, this approach has been tested in the past and it was much less performant than matching exactly on the event name, as we have now.

Agreed that performance is critical here. I experimented with this using `ets:match` instead of `ets:lookup` and it is slower:

```
Name             ips        average  deviation         median         99th %
lookup        2.87 M        0.35 μs  ±9375.10%           0 μs           1 μs
match         0.42 M        2.39 μs   ±567.89%           2 μs           5 μs
```

```elixir
:ets.new(:table, [:duplicate_bag, :named_table])
:ets.insert(:table, {[:one], 1})
:ets.insert(:table, {[:one, :two], 12})
:ets.insert(:table, {[:one, :two, :three], 123})
:ets.insert(:table, {[:one, :two, :three, :four], 1234})

for n <- 1..20 do
  :ets.insert(:table, {[:something, :else, n], n})
end

Benchee.run(%{
  "match" => fn -> :ets.match(:table, {[:one, :two | :_], :"$1"}) end,
  "lookup" => fn -> :ets.lookup(:table, [:one, :two, :three]) end
})
```

**Comment by hauleth** [comment]
For me the "API" should be as simple as:

```erlang
% Spans are defined as (list of) tuples in form of `{StartEvent, StopEvent}`
-telemetry_span({[repo, query, start], [repo, query, stop]}).

% Metrics are defined as a list of tuples `{Event, Description, MeasurementsKey, MetadataKeys}`
-telemetry_events([{[repo, query, start], "fired on query start", [], [repo, table]},
                   {[repo, query, stop], "fired on query finish", [idle_time, queue_time, query_time, decode_time], [repo, table]}]).
```

In that form we have all informations there and user can fetch required informations with ease via `Module:module_info/1`. Of course it will fail on stripped modules, but that is the price for simple API.

**Comment by bryannaegele** [comment]
I think we could simplify span definitions with events to just the event minus the suffix since we've settled on `start` and `stop` as the blessed way.

```erlang
-telemetry_span([repo, query]).
```

@hauleth I added an event data structure definition item to the proposal list for discussion. 

For the general discussion, I think all of the items in the proposal could potentially be wrapped up the idea of a `telemetry_registry` which could just be a separate experimental library which we could vet some of these ideas and usage prior to potentially adding any of this to the core library. I think proving out these ideas for discovery and definition outside of the main library with a goal of possibly merging them into the core library for the 1.0 release would be prudent given the proliferation of the core library.

**Comment by hauleth** [comment]
> I think we could simplify span definitions with events to just the event minus the suffix since we've settled on start and stop as the blessed way.

The problems I see:

- It is only "blessing" with no (even partially) enforced strategy. There is no functions like `start_span`, `finish_span`, or `with_span` that would at least partially enforce such naming.
- It assumes that each module will use only one prefix. Imagine that we would have `[repo, query, idle]`, `[repo, query, queue]`, `[repo, query, query]`, and `[repo, query, decode]` spans. With your approach there is not really simple way to define all of them.

**Comment by bryannaegele** [comment]
We wouldn't need enforcement of naming if we are adding `start` and `stop` to the handlers from the definition. From the definition `-telemetry_span([repo, query]).` you'd create handlers on `[repo, query, start]` and `[repo, query, stop]`. Or maybe I'm missing something.

**Comment by bryannaegele** [comment]
To the second point, each span event name has to be declared and given start / stop events. I'm not sure what the events you're describing are. My approach was an attempt at preventing the need to list a start and a stop event separately for every span. That feels cumbersome to me.

**Comment by hauleth** [comment]
I would say that we should keep `{StartEvent,StopEvent}` with eventual macro that would generate such entry from the prefix alone. Ex.:

```erlang
-define(TELEMETRY_SPAN(Prefix), -telemetery_span({[Prefix | [start]], [Prefix | [stop]]})).

?TELEMETRY_SPAN([repo, query]).
```

Only thing needed afterwards is to flatten the spans when fetching. Unfortunately there is no way to do `Prefix ++ [start]` in Erlang attributes.

**Comment by hauleth** [comment]
IMHO it would be also nice to have `telemetry:span/2` function that would "fire" these events for user "automagically". It a little bit diverges from the [Clever Functional Design](https://ferd.ca/clever-functional-design.html), but IMHO provide clear approach (and also will partially ~~force~~ mildly suggest users to use `start` and `stop` events on all spans).

**Comment by josevalim** [comment]
There is a separate discussion for span where we talked about this but the
consensus is that you often pass different metadata to start and stop and
providing a concise API that handles all of the possible scenarios is
easier said than done.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by hauleth** [comment]
@josevalim my point was to make "simple" cases simpler and giving user suggestion that we prefer such suffixes. Not to force users to use such function or to prevent sending such events independently. Just as a nice convenience function that "shows" suggested approach.

**Comment by josevalim** [comment]
So can we please move this discussion to a separate issue or to its
original issue on telemetry_metrics? I don’t think there is such a simple
API if you want to dispatch different events in start/stop, which is 99% of
the cases, but I am glad to be proven wrong as long as we don’t side track
this discussion.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by chasers** [comment]
> We had this discussion a couple times already, so we should stick the discussion somewhere so we don't revisit it every 2 or 3 months. There are reasons why telemetry works better than logger for what it intends to but I don't think we should side-track this issue by rehashing the same arguments.

@josevalim would love to learn more about this if you can find the previous discussion?

I hope to help around here once I am able to dive into Telemetry more.

**Comment by bryannaegele** [comment]
@chasers if you search `logger` in this repo's issues or the possibly the org, you should find some discussions. https://github.com/beam-telemetry/telemetry/issues?utf8=%E2%9C%93&q=is%3Aissue+logger 

We've also had some during WG meetings but we don't have recordings of those yet. We're looking into recording our meetings going forward.

**Comment by josevalim** [comment]
@bryannaegele I believe we have a lot of different experiments on this, including telemetry_registry. Shall we close this issue? Or shall we start a new issue with a more up to date discussion? Most of the discussion here is stale. :)

**Comment by bryannaegele** [comment]
Yeah. I think at this point that this functionality would just live in the separate library. I'll close this and create new issues in the other repo when I get some time.
